src/revocation.rs

```diff
@@ -0,0 +1,131 @@
+    let revocation_cert_path =
+        format!("{}/unzipped/RevocationNotifier-cert.crt", mount);
+
+    // Connect to the service via 0mq
+    let context = zmq::Context::new();
+    let mysock = context.socket(zmq::SUB)?;
+
+    mysock.set_subscribe(b"")?;
+
+    let revocation_ip =
+        config_get("/etc/keylime.conf", "general", "receive_revocation_ip")?;
+    let revocation_port = config_get(
+        "/etc/keylime.conf",
+        "general",
+        "receive_revocation_port",
+    )?;
+    let endpoint = format!("{}:{}", revocation_ip, revocation_port);
+
+    mysock.connect(endpoint.as_str())?;
+
+    // Unlike the python agent we do not attempt lazy loading. We either
+    // have the certificate, or we don't. If we don't have a key or can't load
+    // the key we return a Configuration error as the service will not work.
+    let cert_key = if Path::new(&revocation_cert_path).exists() {
+        info!(
+            "Loading the revocation certificate from {}",
+            revocation_cert_path
+        );
+        match crypto::rsa_import_pubkey(revocation_cert_path.clone()) {
+            Ok(v) => v,
+            Err(e) => {
+                return Err(error::Error::Configuration(String::from(
+                    "Can not load pubkey",
+                )))
+            }
+        }
+    } else {
+        error!("Path for the 0mq socket socket doesn't exist");
+        return Err(error::Error::Configuration(String::from(
+            "Path for the 0mq socket socket doesn't exist",
+        )));
+    };
+
+    info!("Waiting for revocation messages on 0mq {}", endpoint);
+
+    // Main revocation service loop. If a message is malformed or
+    // can not be verified the loop continues.
+    loop {
+        let mut rawbody = match mysock.recv_string(0) {
+            Ok(v) => match v {
+                Ok(v) => v,
+                _ => {
+                    warn!("Unable to read message from 0mq");
+                    continue;
+                }
+            },
+            Err(e) => {
+                warn!("Unable to read message from 0mq");
+                continue;
+            }
+        };
+
+        let body: Value = serde_json::from_str(rawbody.as_str())?;
+
+        // Ensure we have a signature, otherwise continue the loop
+        let signature = match body.get(&"signature") {
+            Some(v) => v.to_string(),
+            _ => {
+                warn!("No signature on revocation message from server");
+                continue;
+            }
+        };
+
+        // Ensure we have a msg, otherwise continue the loop
+        let message = match body["msg"].as_str() {
+            Some(v) => v,
+            _ => {
+                warn!("No msg on revocation message from server");
+                continue;
+            }
+        };
+
+        // Verify the message and signature with our key
+        let mut verified =
+            crypto::rsa_verify(cert_key.clone(), message, signature);
```

Comment:
Is there any chance `cert_key` could be given as a reference too (`&...`)? That would save us from having to `clone()` every time we get a message.

Reply:
We can make `rsa_verify` take a reference but once inside the function it seems like we need to clone for [`PKey::from_rsa`](https://docs.rs/openssl/0.10.31/openssl/pkey/struct.PKey.html#method.from_rsa) though I may be wrong. Thoughts?

Reply:
Bump :-)

Reply:
I am good with this now, @puiterwijk if @ashcrow answer satisfies you, we can merge this

Reply:
Sure. Let's merge it like this and improve later.